File: 2. Insert at beginning.cpp
#include <iostream>
using namespace std;

struct Node { int data; Node* next; Node(int d):data(d),next(nullptr){} };

void push_front(Node*& head, int val){ Node* n=new Node(val); n->next=head; head=n; }
void print(Node* head){ while(head){ cout<<head->data; if(head->next) cout<<" -> "; head=head->next;} cout<<endl; }

int main(){ Node* head=nullptr; push_front(head,30); push_front(head,20); push_front(head,10); cout<<"After inserting at beginning: "; print(head); return 0; }

File: 2. Insert at beginning.py
"""
Task: Insert nodes at the beginning of a singly linked list
Time Complexity: O(1) per insert at front
Space Complexity: O(n)
"""


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def push_front(self, val):
        n = Node(val)
        n.next = self.head
        self.head = n

    def print_list(self):
        cur = self.head
        out = []
        while cur:
            out.append(str(cur.data))
            cur = cur.next
        print(' -> '.join(out))


if __name__ == '__main__':
    ll = LinkedList()
    ll.push_front(30)
    ll.push_front(20)
    ll.push_front(10)
    print('After inserting at beginning: ', end='')
    ll.print_list()

# Task: Insert nodes at the beginning of a singly linked list
